### PR TITLE
feat(deprecation): add `__qualname__` to deprecated args warnings

### LINKED ISSUE #2010: Unclear which `HTTPError` was instantiated from `deprecated_args()` warnings
In Falcon 3.0.x, instantiating `HTTPError` or its subclasses with positional arguments (of course except the allowed ones) generates a `DeprecatedWarning` via the `deprecated_args()` decorator.

However, it is unclear from the warning which class/function was invoked, it just says "calls [with more than N] positional args are deprecated". Brought up by @laurent-chriqui (see the linked PR).

Ideally, as a developer, I would like the warning to read along the lines of
```
DeprecatedWarning: Calls to HTTPNotFound.__init__(...) with positional args are deprecated. Please specify them as keyword arguments instead.
```

### PR DESCRIPTION
**Pull request by laurent-chriqui**
# Summary of Changes

The deprecated args warning does not tell the function name which is then very hard to debug.

# Related Issues

Closes #2010 

# Pull Request Checklist

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [X] Applied changes to both WSGI and ASGI code paths and interfaces (where applicable).
- [X] Added **tests** for changed code.
- [X] Prefixed code comments with GitHub nick and an appropriate prefix.
- [X] Coding style is consistent with the rest of the framework.
- [X] Updated **documentation** for changed code.
    - [X] Added docstrings for any new classes, functions, or modules.
    - [X] Updated docstrings for any modifications to existing code.
    - [X] Updated both WSGI and ASGI docs (where applicable).
    - [X] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [X] Updated all relevant supporting documentation files under `docs/`.
    - [X] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [X] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [X] Changes (and possible deprecations) have [towncrier](https://towncrier.readthedocs.io/en/actual-freaking-docs/index.html) news fragments under `docs/_newsfragments/`, with the file name format `{issue_number}.{fragment_type}.rst`. (Run `towncrier --draft` to ensure it renders correctly.)

If you have *any* questions to *any* of the points above, just **submit and ask**! This checklist is here to *help* you, not to deter you from contributing!

### CODE REVIEW

**Review by sde-melo — DISMISSED** [review]

**Review by vytas7 — CHANGES_REQUESTED** [review]
Hi @laurent-chriqui, and thanks for this improvement!

I agree that the way these warnings are emitted at the time of this writing may be hard for the user to track down.
However, even after your changes, it won't probably become much clearer either for the most important use case of `deprecated_args` within the framework itself: [`HTTPError.__init__(...)`](https://falcon.readthedocs.io/en/stable/api/errors.html#falcon.HTTPError):
```python
>>> from falcon import HTTPNotFound
>>> HTTPNotFound('I looked everywhere...')
<stdin>:1: DeprecatedWarning: Calls to __init__ with positional args are deprecated. Please specify them as keyword arguments instead.
<HTTPNotFound: 404 Not Found>
```

Maybe we could add something along the lines of `method_name` I added to the `deprecated()` decorator above? Or some other machinery to infer the class's name?

Ideally, as a developer, I would like to get something like:
```
DeprecatedWarning: Calls to HTTPNotFound.__init__(...) with positional args are deprecated. Please specify them as keyword arguments instead.
```

In addition, it would be cool to add a towncrier newsfragment for this improvement (of type `.newandimproved.rst`).

**Comment by laurent-chriqui** [comment]
@vytas7 Thank you for pointing that out. I will make the necessary changes right away.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Maybe this got a bit overly complex now... I understand the value proposition, but usually similar information (such as the file name and source code line) is automatically emitted by the interpreter alongside a warning.

Having too complex logic in such a place also increases the possibility of unforeseen bugs (e.g., https://github.com/falconry/falcon/issues/1882; taking `__repr__` of arbitrary stuff might also have unforeseen side effects if ppl pass some weird stuff in)... So bonus points if we could somehow simplify this to `ClassName.method_name(...)` if we're dealing with a class method, and just `method_name(...)` otherwise.

All that being said, thanks once again for this PR. This is a great usability improvement for our users, and it would be actually nice to add a Towncrier newsfragment reflecting this.

**Comment by laurent-chriqui** [comment]
It can be quite simple if we only want the Class name and method/function name without the passed arguments.
When you write `ClassName.method_name(...)` what do you mean by the three dots ? I thought you meant that you wanted the arguments to be repeated but if you mean them litterally, we just need to pull  `fn.__qualname__` and it will be simple.

**Comment by vytas7** [comment]
Is `__qualname__` always available for, e.g., cythonized methods?

**Comment by laurent-chriqui** [comment]
According to my tests, `__qualname__` does work on compiled code.

File: tests/test_utils.py
Comment (vytas7) [inline tests/test_utils.py:1324]: Could we change this assertion to ensure that `C.` from `__qualname__` is surfaced in the message as well?

Comment (laurent-chriqui) [reply tests/test_utils.py]: done

File: docs/_newsfragments/2010.newandimproved.rst
Comment (vytas7) [inline docs/_newsfragments/2010.newandimproved.rst:3]: Suggested edit:
```rst
When called with deprecated positional arguments, methods and class
initializers (such as :class:`falcon.errors.HTTPError`) will now emit a
user-friendlier warning indicating the fully qualified name of the method in
question.
```

**Review by vytas7 — CHANGES_REQUESTED** [review]
LGTM now :+1: 

Although not critical, I would just like to polish the newsfragment somewhat (and since you didn't allow maintainer edits in this PR, it's only you who can make the update).
Please check if my suggestion makes sense and/or improve it if my version sounds funky to you :smiling_imp:

**Review by CaselIT — APPROVED** [review]
lgtm

**Comment by kgriffs** [comment]
LGTM modulo the news fragment.

File: docs/_newsfragments/2010.newandimproved.rst
Comment (laurent-chriqui) [reply docs/_newsfragments/2010.newandimproved.rst]: done

**Review by vytas7 — APPROVED** [review]
Thanks for this improvement! :tada:

**Pull request merged**
